### PR TITLE
`mul_shr`implementation and coverage

### LINKED ISSUE #16: [Feature]: `mul_shr` function
Implement a function that calculates `floor(x * y >> n)` with full precision.

Check out [Solidity Math Library](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/utils/math/Math.sol) as a reference.

### PR DESCRIPTION
**Pull request by ericnordelo**
Resolves #16 

#### PR Checklist

- [x] Tests
- [ ] Documentation
- [ ] Changelog

## Release Notes

* **New Features**
  * Introduced multiply-shift-right operation with rounding support across all integer types.
  * Enhanced multiply-divide operations with improved overflow detection and return values.
  * Added comprehensive rounding modes for arithmetic operations.

* **Tests**
  * Expanded test coverage for arithmetic operations, edge cases, and rounding behavior.

### CODE REVIEW

**Review by bidzyyys — APPROVED** [review]
Looks good!

File: math/core/sources/macros.move [hunk 1]
```diff
@@ -221,38 +266,143 @@ public(package) fun mul_div_u256_wide(
...
+/// and `result` contains the shifted and rounded value when no overflow occurred.
+public(package) fun mul_shr_u256_wide(
+    a: u256,
+    b: u256,
+    shift: u8,
+    rounding_mode: RoundingMode,
+): (bool, u256) {
+    let product = u512::mul_u256(a, b);
+    let hi = u512::hi(&product);
+    let lo = u512::lo(&product);
```
Comment (immrsd) [inline math/core/sources/macros.move]:
```suggestion
    let hi = product.hi();
    let lo = product.lo();
```

File: math/core/sources/macros.move [hunk 2]
```diff
@@ -221,38 +266,143 @@ public(package) fun mul_div_u256_wide(
...
-        if (should_round_up) {
-            if (quotient == std::u256::max_value!()) {
-                return (true, 0)
-            };
-            quotient = quotient + 1;
-        }
+        return round_division_result(quotient, denominator, remainder, rounding_mode)
     };
 
     (false, quotient)
```
Comment (immrsd) [inline math/core/sources/macros.move]: We might write it using if-else and go without early returns here
```
if (overflow) {
    (true, 0)
} else if (remainder == 0) {
    (false, quotient)
} else {
    round_division_result(quotient, denominator, remainder, rounding_mode)
}
```

File: math/core/sources/macros.move [hunk 3]
```diff
@@ -138,8 +138,61 @@ public(package) macro fun mul_div<$Int>(
...
+///
+/// #### Aborts
+/// Does not emit custom errors, but will inherit the Move abort that occurs when `$shift` is 256 or
+/// greater.
+public(package) macro fun mul_shr<$Int>(
+    $a: $Int,
+    $b: $Int,
+    $shift: u8,
+    $rounding_mode: RoundingMode,
+): (bool, u256) {
```
[the remark this answers is not in the document]
Comment (ericnordelo) [reply math/core/sources/macros.move]: I don't think we have to/should. As long as the macros are internal to the package, the bool for overflow is more descriptive, and also more efficient gas wise than using Option (which is a struct). We should be ok with keeping option only in the public API unless it makes sense to use it semantically.
